pkg/terraform/terraform.go

```diff
@@ -0,0 +1,190 @@
+}
+
+// Validate runs the terraform validate command.
+func (t *TerraformClient) Validate(ctx context.Context, workingDir string, args ...string) ([]byte, []byte, int, error) {
+	childArgs := []string{"validate"}
+	childArgs = append(childArgs, args...)
+	return child.Run(ctx, workingDir, "terraform", childArgs) //nolint:wrapcheck
+}
+
+// Show runs the Terraform show command.
+func (t *TerraformClient) Show(ctx context.Context, workingDir, file string, args ...string) ([]byte, []byte, int, error) {
+	childArgs := []string{"show"}
+	childArgs = append(childArgs, args...)
+
+	if file != "" {
+		childArgs = append(childArgs, file)
+	}
+
+	return child.Run(ctx, workingDir, "terraform", childArgs) //nolint:wrapcheck
+}
+
+// Plan runs the Terraform plan command.
+func (t *TerraformClient) Plan(ctx context.Context, workingDir, file string, args ...string) ([]byte, []byte, int, error) {
+	childArgs := []string{"plan"}
+	childArgs = append(childArgs, args...)
+
+	if file != "" {
+		childArgs = append(childArgs, fmt.Sprintf("-out=%s", file))
+	}
+
+	return child.Run(ctx, workingDir, "terraform", childArgs) //nolint:wrapcheck
+}
+
+// Apply runs the Terraform apply command.
+func (t *TerraformClient) Apply(ctx context.Context, workingDir, file string, args ...string) ([]byte, []byte, int, error) {
+	childArgs := []string{"apply"}
+	childArgs = append(childArgs, args...)
+
+	if file != "" {
+		childArgs = append(childArgs, file)
+	}
+
+	return child.Run(ctx, workingDir, "terraform", childArgs) //nolint:wrapcheck
+}
+
+// GetEntrypointDirectories gets all the directories that have Terraform config
+// files containing a backend block to be used as an entrypoint module.
+func (t *TerraformClient) GetEntrypointDirectories(rootDir string) ([]string, error) {
+	matches := make(map[string]struct{})
+	if err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to walk directory %s: %w", path, err)
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		matched, err := filepath.Match("*.tf", filepath.Base(path))
+		if err != nil {
+			return fmt.Errorf("failed to find terraform files: %w", err)
+		}
+
+		if !matched {
+			return nil
+		}
+
+		hasBackend, _, err := hasBackendConfig(path)
+		if err != nil {
+			return fmt.Errorf("failed to find terraform backend config: %w", err)
+		}
+
+		if !hasBackend {
+			return nil
+		}
+
+		matches[filepath.Dir(path)] = struct{}{}
+
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to find files: %w", err)
+	}
+
+	dirs := []string{}
+
+	for dir := range matches {
+		dirs = append(dirs, dir)
+	}
+
+	sort.Strings(dirs)
+
+	return dirs, nil
+}
+
+// hasBackendConfig tests a Terraform config file for the existence of a backend block.
+func hasBackendConfig(path string) (bool, hcl.Diagnostics, error) {
+	var diags hcl.Diagnostics
+
+	if _, err := os.Stat(path); err != nil {
```

Comment:
This Stat() call is arguably redundant. We should be checking the result of ParseHCLFile anyway, which would include the case for nonexistent file.

https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclparse#Parser.ParseHCLFile says `An error diagnostic is returned if the given file cannot be read.` Check out https://github.com/hashicorp/hcl/blob/v2.17.0/hclparse/parser.go#L78 to see what the error looks like.